content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,563 @@
+
+```python
+class EmbeddingVectorizer:
+    def __init__(self):
+        """
+        Embedding vectorizer
+        """
+        fasttext.util.download_model('fr', if_exists='ignore')
+        self.embedding_model = fasttext.load_model("./cc.fr.300.bin")
+
+    def __call__(self, address):
+        """
+        Convert address to embedding vectors
+        :param address: The address to convert
+        :return: The embeddings vectors
+        """
+        embeddings = []
+        for word in address.split():
+            embeddings.append(self.embedding_model[word])
+        return embeddings
+     
+embedding_vectorizer = EmbeddingVectorizer()
+```
+
+We also need to apply a similar operation to the address tags (e.g. StreeNumber, StreetName). 
+This time, however, the vectorizer needs to convert the tags into categorical values (e.g. 0, 1). 
+
+For simplicity, we will use a `DatasetVectorizer` class that will apply the vectorizing process using both 
+the embedding and the address vectorize process.
+
+```python
+class DatasetVectorizer:
+    def __init__(self, embedding_vectorizer):
+        self.embedding_vectorizer = embedding_vectorizer
+        self.tags_set = {
+            "StreetNumber": 0,
+            "StreetName": 1,
+            "Unit": 2,
+            "Municipality": 3,
+            "Province": 4,
+            "PostalCode": 5,
+            "Orientation": 6,
+            "GeneralDelivery": 7
+        }
+
+    def vectorize(self, data):  # We vectorize inplace
+        for idx, item in enumerate(data):
+            data[idx] = self._item_vectorizing(item)
+
+    def _item_vectorizing(self, item):
+        address = item[0]
+        address_vector = self.embedding_vectorizer(address)
+
+        tags = item[1]
+        idx_tags = self._convert_tags_to_idx(tags)
+
+        return address_vector, idx_tags
+
+    def _convert_tags_to_idx(self, tags):
+        idx_tags = []
+        for tag in tags:
+            idx_tags.append(self.tags_set[tag])
+        return idx_tags
+
+dataset_vectorizer = DatasetVectorizer(embedding_vectorizer)
+```  
+
+Now, let's vectorize (in place) our dataset.
+
+```python
+dataset_vectorizer.vectorize(train_data)
+dataset_vectorizer.vectorize(valid_data)
+dataset_vectorizer.vectorize(test_data)
+```
+
+### DataLoader
+> We use a first trick, ``padding``.
+
+Now, since all the addresses are not of the same size, it is impossible to batch them together since all tensor elements must have the same lengths. But there is a trick, padding!
+
+The idea is simple. We add *empty* tokens at the end of each sequence up to the longest one in a batch. For the word vectors, we add vectors of 0 as padding. For the tag indices, we pad with -100s. We do so because of the [cross-entropy loss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) and the accuracy metric all ignore targets with values of -100.
```

Comment:
```suggestion
The idea is simple. We add *empty* tokens at the end of each sequence up to the longest one in a batch. For the word vectors, we add vectors of 0 as padding. For the tag indices, we pad with -100's. We do so because the [cross-entropy loss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) and the accuracy metric both ignore targets with values of -100.
```